glide-core/redis-rs/redis/src/cluster_async/pipeline_routing.rs

```diff
@@ -0,0 +1,519 @@
+///
+/// It also collects response policies for multi-node routing and returns them along with the pipeline map.
+/// This is to ensure we can aggregate responses from properly from the different nodes.
+///
+/// # Arguments
+///
+/// * `pipeline` - A reference to the pipeline containing the commands to route.
+/// * `core` - The core object that provides access to connection locks and other resources.
+///
+/// # Returns
+///
+/// A `Result` containing a tuple:
+/// - A `NodePipelineMap<C>` where commands are grouped by their corresponding nodes (as pipelines).
+/// - A `Vec<(usize, MultipleNodeRoutingInfo, Option<ResponsePolicy>)>` containing the routing information
+///   and response policies for multi-node commands, along with the index of the command in the pipeline, for aggregating the responses later.
+pub async fn map_pipeline_to_nodes<C>(
+    pipeline: &crate::Pipeline,
+    core: Core<C>,
+) -> Result<
+    (
+        NodePipelineMap<C>,
+        Vec<(usize, MultipleNodeRoutingInfo, Option<ResponsePolicy>)>,
+    ),
+    (OperationTarget, RedisError),
+>
+where
+    C: Clone + ConnectionLike + Connect + Send + Sync + 'static,
+{
+    let mut pipelines_per_node = NodePipelineMap::new();
+    let mut response_policies = Vec::new();
+
+    for (index, cmd) in pipeline.cmd_iter().enumerate() {
+        match RoutingInfo::for_routable(cmd).unwrap_or(SingleNode(SingleNodeRoutingInfo::Random)) {
+            SingleNode(route) => {
+                handle_pipeline_single_node_routing(
+                    &mut pipelines_per_node,
+                    cmd.clone(),
+                    route.into(),
+                    core.clone(),
+                    index,
+                )
+                .await?;
+            }
+            MultiNode((multi_node_routing, response_policy)) => {
+                //save the routing info and response policy, so we will be able to aggregate the results later
+                response_policies.push((index, multi_node_routing.clone(), response_policy));
+                match multi_node_routing {
+                    MultipleNodeRoutingInfo::AllNodes | MultipleNodeRoutingInfo::AllMasters => {
+                        let connections: Vec<_> = {
+                            let lock = core.conn_lock.read().expect(MUTEX_READ_ERR);
+                            if matches!(multi_node_routing, MultipleNodeRoutingInfo::AllNodes) {
+                                lock.all_node_connections().collect()
+                            } else {
+                                lock.all_primary_connections().collect()
+                            }
+                        };
+
+                        if connections.is_empty() {
+                            return Err((
+                                OperationTarget::NotFound,
+                                RedisError::from((
+                                    ErrorKind::AllConnectionsUnavailable, // should use different kind? ConnectionNotFoundForRoute
```

Comment:
I think that if we don't have any primary connections it is ok to set the client with AllConnectionsUnavailable

Reply:
just make the error message clearer if it's due to all connections or only primary ones